src/test/java/echoserver/EchoServerTest.java

```diff
@@ -3,36 +3,37 @@
 import org.junit.jupiter.api.Test;
 
 import java.io.*;
-import java.util.ArrayList;
+import java.util.concurrent.Executors;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
 
 class EchoServerTest {
-    void addFakeClientSocketsToList(ArrayList<FakeClientSocket> fakeClientSockets, OutputStream output) {
-        var input1 = new ByteArrayInputStream("Client 1's message".getBytes());
-        var input2 = new ByteArrayInputStream("Client 2's message".getBytes());
-        var input3 = new ByteArrayInputStream("Client 3's message".getBytes());
-        fakeClientSockets.add(new FakeClientSocket(input1, output));
-        fakeClientSockets.add(new FakeClientSocket(input2, output));
-        fakeClientSockets.add(new FakeClientSocket(input3, output));
-    }
-
-
     @Test
-    void connectsToThreeClientsConcurrently() throws IOException {
-        var expectedOutput = "Echo from server: Client 1's message\n" +
-                "Echo from server: Client 2's message\n" +
-                "Echo from server: Client 3's message";
+    void returnsInputOfFourDifferentClients() throws IOException, InterruptedException {
+        var expectedConnectedClients = 4;
+        var expectedOutput1 = "Echo from server: Client 1's message\n";
+        var expectedOutput2 = "Echo from server: Client 2's message\n";
+        var expectedOutput3 = "Echo from server: Client 3's message\n";
+        var expectedOutput4 = "Echo from server: Client 4's message\n";
         var output = new ByteArrayOutputStream();
-
-        var fakeClientSockets = new ArrayList<FakeClientSocket>();
-        addFakeClientSocketsToList(fakeClientSockets, output);
-        var serverSocket = new FakeServerSocket(fakeClientSockets);
-        var echoServer = new EchoServer(serverSocket);
-
-        echoServer.start();
-
-
-        assertThat(output.toString()).contains(expectedOutput);
+        var fakeServerSocket = new FakeServerSocket(output);
+        fakeServerSocket.setupFakeClientSockets();
+        var executor = Executors.newFixedThreadPool(4);
+        var echoServer = new EchoServer(fakeServerSocket, executor);
+
+        echoServer.listenForClients();
+        echoServer.listenForClients();
+        echoServer.listenForClients();
+        echoServer.listenForClients();
+        Thread.sleep(100);
```

Comment:
Why sleep and why 100ms?

Reply:
For some reason, when `Thread.sleep` is removed, this test fails (and passes sometimes). It appears all or some of the output streams are blank, as if the test ends prematurely before all output is received. I'm sure I'm missing something. Do you know why this may be occurring?

In regards to the milliseconds, after trying `1000`, `100` didn't cause a noticeable pause when running the test.

Reply:
I think this is related to `Executors` running tasks in an asynchronous way - it's worth looking into it, especially to check if there is a way to make it synchronous for the tests (so that you wouldn't have to use `Thread.sleep(100)`).